File: backend/schemas/schedule.py
"""Schedule grid API schemas."""
from pydantic import BaseModel, ConfigDict, Field, field_validator

from backend.schemas.common import TeacherBrief


class ShiftBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    school_level: str
    working_days: int
    max_lessons_per_day: int
    start_lesson: int
    lessons_count: int
    class_hour_day: int | None = None
    class_hour_time_label: str | None = None
    class_hour_lessons_count: int | None = None


class SchoolClassRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    grade: int
    school_level: str
    shift_id: int | None = None
    home_classroom_id: int | None = None


class ScheduleCellOut(BaseModel):
    id: int
    class_id: int
    day_of_week: int
    lesson_number: int
    assignment_id: int
    classroom_id: int | None = None
    subject_id: int
    subject_name: str
    subject_color: str
    teacher_id: int | None = None
    teacher_name: str | None = None
    group_number: int | None = None
    classroom_name: str | None = None
    requires_fixed_classroom: bool = False


class ScheduleSettingsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    school_level: str
    max_lessons_per_subject_per_day: int
    classroom_mode: str
    elementary_group_subjects_leave: bool
    pref_teacher_gaps: int = 5
    pref_hard_subjects_early: int = 5
    pref_adjacent_pairs: int = 5
    pref_classroom_stability: int = 5


class ClassroomWarningOut(BaseModel):
    type: str
    message: str


class ScheduleGridOut(BaseModel):
    school_level: str
    current_shift_id: int | None = None
    current_shift: ShiftBrief | None = None
    shifts: list[ShiftBrief]
    classes: list[SchoolClassRow]
    day_names: list[str]
    working_days: int
    max_lessons: int
    lessons_range: list[int]
    lesson_times_by_day: dict[int, dict[int, str]]
    class_hour_time_label: str
    cells: list[ScheduleCellOut]
    classroom_warnings: list[ClassroomWarningOut]
    settings: ScheduleSettingsOut | None = None


class AssignmentChoiceOut(BaseModel):
    id: int
    subject_id: int
    subject_name: str
    subject_color: str
    teacher_id: int | None = None
    teacher_name: str | None = None
    group_number: int | None = None
    remaining_hours: int
    preferred_classroom_id: int | None = None
    requires_fixed_classroom: bool = False


class ClassroomChoiceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    number: str
    name: str | None = None
    display_name: str
    subject_ids: list[int] = []
    is_exclusive: bool = False
    school_level: str | None = None
    subgroup_only: bool = False
    classes_capacity: int = 1


class AssignmentsForClassOut(BaseModel):
    assignments: list[AssignmentChoiceOut]
    classrooms: list[ClassroomChoiceOut]


class TeacherDayOccupantOut(BaseModel):
    class_id: int
    class_name: str
    subject_name: str
    subject_color: str
    classroom_name: str | None = None
    group_number: int | None = None


class TeacherDayLessonOut(BaseModel):
    lesson: int
    time_label: str | None = None
    is_candidate: bool = False
    is_gap: bool = False
    overlaps_current: bool = False
    occupants: list[TeacherDayOccupantOut] = []


class TeacherDayShiftOut(BaseModel):
    shift_id: int | None = None
    shift_name: str
    is_current: bool = False
    lessons: list[TeacherDayLessonOut]


class TeacherDayOut(BaseModel):
    teacher_id: int
    teacher_name: str
    day_of_week: int
    day_name: str
    other_shift_gap: str | None = None
    shifts: list[TeacherDayShiftOut]


class ScheduleCellCreate(BaseModel):
    class_id: int
    day_of_week: int = Field(..., ge=1, le=6)
    lesson_number: int = Field(..., ge=0, le=20)
    assignment_id: int
    classroom_id: int | None = None


class ScheduleCellMove(BaseModel):
    day_of_week: int = Field(..., ge=1, le=6)
    lesson_number: int = Field(..., ge=0, le=20)
    class_id: int | None = None
    classroom_id: int | None = None
    set_classroom: bool = False


class ScheduleCellSwapClassroom(BaseModel):
    other_cell_id: int


class ScheduleCellSwapOut(BaseModel):
    cell: ScheduleCellOut
    other: ScheduleCellOut


class AutoPageData(BaseModel):
    teachers: list[TeacherBrief]
    classes: list[SchoolClassRow]
    elementary_warnings: list[ClassroomWarningOut]
    secondary_warnings: list[ClassroomWarningOut]
    elementary_settings: ScheduleSettingsOut | None = None
    secondary_settings: ScheduleSettingsOut | None = None
    shifts_elementary: list[ShiftBrief]
    shifts_secondary: list[ShiftBrief]


class AutoAllStreamBody(BaseModel):
    school_level: str = Field("elementary", pattern="^(elementary|secondary)$")
    shift_id: int
    time_limit_sec: float = Field(60.0, ge=1, le=86_400)
    random_seed: int = 1
    diagnose: bool = False
    split: str = Field("shift", pattern="^(shift|grade_bands)$")
    hours_first: str = Field("more", pattern="^(more|fewer)$")


class AutoByTeacherStreamBody(BaseModel):
    teacher_id: int
    school_level: str = Field("elementary", pattern="^(elementary|secondary)$")
    diagnose: bool = False


class ClearScheduleBody(BaseModel):
    school_level: str | None = Field(None, pattern="^(elementary|secondary)$")
    class_id: int | None = None
    teacher_id: int | None = None
    days_of_week: list[int] | None = Field(None, min_length=1)

    @field_validator("days_of_week")
    @classmethod
    def _days_of_week(cls, value: list[int] | None) -> list[int] | None:
        if value is None:
            return None
        seen: set[int] = set()
        days: list[int] = []
        for day in value:
            if day < 1 or day > 6:
                raise ValueError("days_of_week must be 1..6 (Mon..Sat)")
            if day not in seen:
                seen.add(day)
                days.append(day)
        return days


class ClearScheduleResult(BaseModel):
    count: int


class SettingsPair(BaseModel):
    elementary: ScheduleSettingsOut | None = None
    secondary: ScheduleSettingsOut | None = None


class SettingsUpdate(BaseModel):
    max_lessons_per_subject_per_day: int = Field(..., ge=1, le=5)
    classroom_mode: str = Field(..., pattern="^(class_room|teacher_room)$")
    elementary_group_subjects_leave: bool | None = None
    pref_teacher_gaps: int = Field(5, ge=0, le=10)
    pref_hard_subjects_early: int = Field(5, ge=0, le=10)
    pref_adjacent_pairs: int = Field(5, ge=0, le=10)
    pref_classroom_stability: int = Field(5, ge=0, le=10)


class ExplainSlotBody(BaseModel):
    assignment_id: int
    day_of_week: int = Field(..., ge=1, le=6)
    lesson_number: int = Field(..., ge=0, le=20)
    classroom_id: int | None = None
    cell_id: int | None = None


class ExplainSlotOut(BaseModel):
    allowed: bool
    blockers: list[str]
    alternatives: list[dict]
    text: str
    llm_used: bool


class RepairBody(BaseModel):
    school_level: str = Field("elementary", pattern="^(elementary|secondary)$")
    teacher_id: int | None = None
    class_id: int | None = None


class AssistBody(BaseModel):
    message: str = Field(..., min_length=1, max_length=500)
    school_level: str = Field("elementary", pattern="^(elementary|secondary)$")
    shift_id: int | None = None
    apply: bool = False


class AssistMoveOut(BaseModel):
    cell_id: int
    subject: str
    class_name: str
    from_day: int
    from_lesson: int
    to_day: int
    to_lesson: int
    allowed: bool
    applied: bool
    blockers: list[str]
    label: str


class AssistOut(BaseModel):
    interpretation: str
    llm_used: bool
    preference_updates: dict[str, int]
    preferences_applied: bool
    moves: list[AssistMoveOut]
    applied_moves: int
    rejected: list[AssistMoveOut]

File: frontend/src/api/schedule.ts
import { apiJson, extractApiError, ApiError } from './client'
import type { ShiftBrief } from './shifts'
import type { SchoolLevel } from '../domain/schoolLevel'

export type ClassroomMode = 'class_room' | 'teacher_room'

export type ScheduleSettings = {
  school_level: string
  max_lessons_per_subject_per_day: number
  classroom_mode: ClassroomMode
  elementary_group_subjects_leave: boolean
  pref_teacher_gaps: number
  pref_hard_subjects_early: number
  pref_adjacent_pairs: number
  pref_classroom_stability: number
}

export type ClassroomWarning = { type: string; message: string }

export type ScheduleCell = {
  id: number
  class_id: number
  day_of_week: number
  lesson_number: number
  assignment_id: number
  classroom_id: number | null
  subject_id: number
  subject_name: string
  subject_color: string
  teacher_id: number | null
  teacher_name: string | null
  group_number: number | null
  classroom_name: string | null
  requires_fixed_classroom?: boolean
}

export type SchoolClassRow = {
  id: number
  name: string
  grade: number
  school_level: string
  shift_id: number | null
}

export type ShiftBriefWithClassHour = ShiftBrief & {
  class_hour_day: number | null
  class_hour_time_label: string | null
  class_hour_lessons_count: number | null
}

export type GridData = {
  school_level: SchoolLevel
  current_shift_id: number | null
  current_shift: ShiftBriefWithClassHour | null
  shifts: ShiftBriefWithClassHour[]
  classes: SchoolClassRow[]
  day_names: string[]
  working_days: number
  max_lessons: number
  lessons_range: number[]
  lesson_times_by_day: Record<number, Record<number, string>>
  class_hour_time_label: string
  cells: ScheduleCell[]
  classroom_warnings: ClassroomWarning[]
  settings: ScheduleSettings | null
}

export type ClassroomChoice = {
  id: number
  number: string
  name: string | null
  display_name: string
  subject_ids: number[]
  is_exclusive: boolean
  school_level?: string | null
  classes_capacity?: number | null
  subgroup_only?: boolean
}

export type AssignmentChoice = {
  id: number
  subject_id: number
  subject_name: string
  subject_color: string
  teacher_id: number | null
  teacher_name: string | null
  group_number: number | null
  remaining_hours: number
  preferred_classroom_id: number | null
  requires_fixed_classroom?: boolean
}

export type AssignmentsData = {
  assignments: AssignmentChoice[]
  classrooms: ClassroomChoice[]
}

export type ScheduleWarning = ClassroomWarning

export type AutoPageData = {
  teachers: { id: number; full_name: string }[]
  classes: { id: number; name: string; school_level: string }[]
  elementary_warnings: ScheduleWarning[]
  secondary_warnings: ScheduleWarning[]
  elementary_settings: ScheduleSettings | null
  secondary_settings: ScheduleSettings | null
  shifts_elementary: ShiftBrief[]
  shifts_secondary: ShiftBrief[]
}

export type JobOut = {
  id: number
  kind: string
  status: string
  progress: { current?: number; total?: number; message?: string } | null
  result: { type?: string; count?: number; message?: string; [k: string]: unknown } | null
  error: string | null
}

export type CreateCellPayload = {
  class_id: number
  day_of_week: number
  lesson_number: number
  assignment_id: number
  classroom_id: number | null
}

export type MoveCellPayload = {
  day_of_week: number
  lesson_number: number
  class_id?: number
  classroom_id?: number | null
  set_classroom?: boolean
}

export type ClearScheduleFilter = {
  school_level?: string
  class_id?: number
  teacher_id?: number
  days_of_week?: number[]
}

export function fetchGrid(schoolLevel: SchoolLevel, shiftId?: number | null) {
  const qs = shiftId ? `&shift_id=${shiftId}` : ''
  return apiJson<GridData>(`/api/schedule/grid?school_level=${schoolLevel}${qs}`)
}

export function createScheduleCell(payload: CreateCellPayload) {
  return apiJson<ScheduleCell>('/api/schedule/cells', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function updateScheduleCell(cellId: number, payload: MoveCellPayload) {
  return apiJson<ScheduleCell>(`/api/schedule/cells/${cellId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  })
}

export function deleteScheduleCell(cellId: number) {
  return apiJson<void>(`/api/schedule/cells/${cellId}`, { method: 'DELETE' })
}

export function swapScheduleClassrooms(cellId: number, otherCellId: number) {
  return apiJson<{ cell: ScheduleCell; other: ScheduleCell }>(
    `/api/schedule/cells/${cellId}/swap-classroom`,
    {
      method: 'POST',
      body: JSON.stringify({ other_cell_id: otherCellId }),
    },
  )
}

export function fetchAssignmentsForClass(
  classId: number,
  slot?: { day: number; lesson: number },
) {
  const q =
    slot != null
      ? `?day_of_week=${slot.day}&lesson_number=${slot.lesson}`
      : ''
  return apiJson<AssignmentsData>(`/api/schedule/assignments-for-class/${classId}${q}`)
}

export type TeacherDayOccupant = {
  class_id: number
  class_name: string
  subject_name: string
  subject_color: string
  classroom_name: string | null
  group_number: number | null
}

export type TeacherDayLesson = {
  lesson: number
  time_label: string | null
  is_candidate: boolean
  is_gap: boolean
  overlaps_current: boolean
  occupants: TeacherDayOccupant[]
}

export type TeacherDayShift = {
  shift_id: number | null
  shift_name: string
  is_current: boolean
  lessons: TeacherDayLesson[]
}

export type TeacherDayData = {
  teacher_id: number
  teacher_name: string
  day_of_week: number
  day_name: string
  other_shift_gap: string | null
  shifts: TeacherDayShift[]
}

export function fetchTeacherDay(params: {
  teacherId: number
  day: number
  classId?: number
  lesson?: number
}) {
  const q = new URLSearchParams()
  q.set('teacher_id', String(params.teacherId))
  q.set('day_of_week', String(params.day))
  if (params.classId != null) q.set('class_id', String(params.classId))
  if (params.lesson != null) q.set('lesson_number', String(params.lesson))
  return apiJson<TeacherDayData>(`/api/schedule/teacher-day?${q.toString()}`)
}

export function fetchAutoPageData() {
  return apiJson<AutoPageData>('/api/schedule/auto/page-data')
}

export function updateScheduleSettings(level: SchoolLevel, payload: Partial<ScheduleSettings>) {
  return apiJson<ScheduleSettings>(`/api/schedule/settings/${level}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  })
}

export function clearSchedule(filter: ClearScheduleFilter) {
  return apiJson<{ count: number }>('/api/schedule/clear', {
    method: 'POST',
    body: JSON.stringify(filter),
  })
}

export function getJob(jobId: number) {
  return apiJson<JobOut>(`/api/jobs/${jobId}`)
}

export function fetchActiveJob() {
  return apiJson<{ job: JobOut | null }>('/api/jobs/active').then((r) => r.job)
}

export function cancelJob(jobId: number, force = false) {
  const q = force ? '?force=true' : ''
  return apiJson<JobOut>(`/api/jobs/${jobId}/cancel${q}`, { method: 'POST' })
}

/** Parse «задача #N» from a 409 conflict body. */
export function stuckJobIdFromError(err: unknown): number | null {
  if (err instanceof ApiError) {
    try {
      const data = JSON.parse(err.body) as { detail?: unknown }
      const detail = data.detail
      if (detail && typeof detail === 'object') {
        const id = (detail as { job_id?: unknown }).job_id
        if (typeof id === 'number' && Number.isFinite(id)) return id
      }
    } catch {
      /* not JSON */
    }
  }
  const msg = extractApiError(err)
  const m = msg.match(/задача #(\d+)/i)
  if (!m) return null
  const id = Number(m[1])
  return Number.isFinite(id) ? id : null
}

export type AutoAllPayload = {
  school_level: SchoolLevel
  shift_id: number
  time_limit_sec: number
  random_seed: number
  diagnose: boolean
  split?: 'shift' | 'grade_bands'
  hours_first?: 'more' | 'fewer'
}

export type AutoByTeacherPayload = {
  teacher_id: number
  school_level: SchoolLevel
  diagnose: boolean
}

export function enqueueAutoAll(payload: AutoAllPayload) {
  return apiJson<{ job_id: number }>('/api/schedule/auto', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export function enqueueAutoByTeacher(payload: AutoByTeacherPayload) {
  return apiJson<{ job_id: number }>('/api/schedule/auto/by-teacher', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export type ExplainSlotPayload = {
  assignment_id: number
  day_of_week: number
  lesson_number: number
  classroom_id?: number | null
  cell_id?: number | null
}

export type ExplainSlotOut = {
  allowed: boolean
  blockers: string[]
  alternatives: {
    day_of_week: number
    lesson_number: number
    day_name: string
    label: string
  }[]
  text: string
  llm_used: boolean
}

export function explainSlot(payload: ExplainSlotPayload) {
  return apiJson<ExplainSlotOut>('/api/schedule/explain', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export type AssistPayload = {
  message: string
  school_level: SchoolLevel
  shift_id?: number | null
  apply?: boolean
}

export type AssistMove = {
  cell_id: number
  subject: string
  class_name: string
  from_day: number
  from_lesson: number
  to_day: number
  to_lesson: number
  allowed: boolean
  applied: boolean
  blockers: string[]
  label: string
}

export type AssistOut = {
  interpretation: string
  llm_used: boolean
  preference_updates: Record<string, number>
  preferences_applied: boolean
  moves: AssistMove[]
  applied_moves: number
  rejected: AssistMove[]
}

export function assistSchedule(payload: AssistPayload) {
  return apiJson<AssistOut>('/api/schedule/assist', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export type RepairPayload = {
  school_level: SchoolLevel
  teacher_id?: number | null
  class_id?: number | null
}

export function enqueueRepair(payload: RepairPayload) {
  return apiJson<{ job_id: number }>('/api/schedule/repair', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export class JobPollAborted extends Error {
  constructor() {
    super('aborted')
    this.name = 'JobPollAborted'
  }
}

function throwIfAborted(signal?: AbortSignal) {
  if (signal?.aborted) throw new JobPollAborted()
}

export async function pollJob(
  jobId: number,
  onProgress: (p: { current: number; total: number; message: string }) => void,
  onLog: (line: string) => void,
  signal?: AbortSignal,
): Promise<JobOut> {
  let lastLogged = ''
  for (;;) {
    throwIfAborted(signal)
    const job = await getJob(jobId)
    throwIfAborted(signal)
    const prog = job.progress || {}
    const message = String(prog.message || '')
    onProgress({
      current: Number(prog.current || 0),
      total: Number(prog.total || 0),
      message: message || job.status,
    })
    if (message && message !== lastLogged) {
      lastLogged = message
      onLog(message)
    }
    if (
      job.status === 'done' ||
      job.status === 'failed' ||
      job.status === 'cancelled'
    ) {
      return job
    }
    await new Promise<void>((resolve, reject) => {
      const onAbort = () => {
        window.clearTimeout(t)
        reject(new JobPollAborted())
      }
      const t = window.setTimeout(() => {
        signal?.removeEventListener('abort', onAbort)
        resolve()
      }, 1000)
      if (!signal) return
      if (signal.aborted) {
        window.clearTimeout(t)
        reject(new JobPollAborted())
        return
      }
      signal.addEventListener('abort', onAbort, { once: true })
    })
  }
}

export async function runJobAndPoll(
  start: () => Promise<{ job_id: number }>,
  onProgress: (p: { current: number; total: number; message: string }) => void,
  onLog: (line: string) => void,
  onStarted?: (jobId: number) => void,
  signal?: AbortSignal,
): Promise<JobOut> {
  const started = await start()
  onStarted?.(started.job_id)
  onLog(`Задача #${started.job_id} поставлена в очередь`)
  return pollJob(started.job_id, onProgress, onLog, signal)
}
